### ISSUE TITLE
Is it possible to define a generic enum ?

### ISSUE THREAD

**Issue by szagi3891** [poem-web/poem#259]
I would like to define a generic enum that I will use in the openapi specification.

I tried to define it this way:

```
use poem_openapi::{OneOf, Object};
use poem_openapi::types::ToJSON;

#[derive(Debug, Object)]
pub struct ProxyResponseOk<R: Send + ToJSON> {
    pub response: R
}

#[derive(Debug, Object)]
pub struct ProxyResponseMessage {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Object)]
pub struct ProxyResponseUnknown {
    pub code: u32,
    pub body: String,
}

#[derive(Debug, OneOf)]
#[oai(property_name="type")]
pub enum ProxyResponse<R: Send + ToJSON> {
    Ok(ProxyResponseOk<R>),
    Error(ProxyResponseMessage),
    Unknown(ProxyResponseUnknown),
}
```

When trying to compile I get errors:

```
error[E0412]: cannot find type `R` in this scope
  --> src/proxy/handler/selection_errors.rs:25:24
   |
22 | #[derive(Debug, OneOf)]
   |                      - help: you might be missing a type parameter: `<R>`
...
25 |     Ok(ProxyResponseOk<R>),
   |                        ^ not found in this scope

error[E0107]: missing generics for enum `ProxyResponse`
  --> src/proxy/handler/selection_errors.rs:24:10
   |
24 | pub enum ProxyResponse<R: Send + ToJSON> {
   |          ^^^^^^^^^^^^^ expected 1 generic argument
   |
note: enum defined here, with 1 generic parameter: `R`
  --> src/proxy/handler/selection_errors.rs:24:10
   |
24 | pub enum ProxyResponse<R: Send + ToJSON> {
   |          ^^^^^^^^^^^^^ -
help: add missing generic argument
   |
24 | pub enum ProxyResponse<R><R: Send + ToJSON> {
   |          ~~~~~~~~~~~~~~~~
```

Can I ask for help?

**Comment by sunli829** [comment]
Currently not supported, I will add it later.

**Comment by sunli829** [comment]
I've added support for generic union in `v1.3.27`, both `inline` and `non-inline`, **`non-inline` requires specifying the name of the concrete type.**

- inline example: https://github.com/poem-web/poem/blob/3efbd920615c7af040169e2f3a07989930809f8f/poem-openapi/tests/union.rs#L452

- non-inline example: https://github.com/poem-web/poem/blob/3efbd920615c7af040169e2f3a07989930809f8f/poem-openapi/tests/union.rs#L467

**Comment by szagi3891** [comment]
There is only one problem:

```
warning: `example-openapi-oneof` (bin "example-openapi-oneof") generated 6 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 1.52s
     Running `target/debug/example-openapi-oneof`
thread 'main' panicked at '`example_openapi_oneof::ProxyResponse<alloc::string::String>` and `example_openapi_oneof::ProxyResponse<u32>` have the same OpenAPI name `ProxyResponse`', /Users/grzegorzszeliga/.cargo/registry/src/github.com-1ecc6299db9ec823/poem-openapi-1.3.27/src/registry/mod.rs:652:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Here is the code that demonstrates this problem

https://github.com/szagi3891/poem-test/blob/master/src/main.rs#L148

**Comment by sunli829** [comment]
If the object is generic, either specify the `inline` attribute to avoid creating the object in the schema. Either use `concrete` to give each concrete type a name.

I'll add a validator to catch this problem at compile time.

**Comment by szagi3891** [comment]
I understand. Would it be possible to make the inline attribute default to a type that is generic?

**Comment by szagi3891** [comment]
```
#[derive(Debug, Object)]
#[oai(inline)]
pub struct ProxyResponseOk<R: Send + ToJSON + ParseFromJSON> {
    pub response: R
}

#[derive(Debug, Object)]
pub struct ProxyResponseMessage {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Object)]
pub struct ProxyResponseUnknown {
    pub code: u32,
    pub body: String,
}

#[derive(Debug, Union)]
#[oai(discriminator_name="type")]
#[oai(inline)]
pub enum ProxyResponse<R: Send + ToJSON + ParseFromJSON> {
    Ok(ProxyResponseOk<R>),
    Error(ProxyResponseMessage),
    Unknown(ProxyResponseUnknown),
}
```

Well, and if inline was added by default for types that have a generic parameter, it should nicely simplify this code.

**Comment by sunli829** [comment]
This is a breaking change 😂, I've improved it now, if you don't specify the `inline` or `concretes` attribute of the generic object, it will report an explicit error at compile time.

**Comment by sunli829** [comment]
Inline objects have the disadvantage that they cannot refer to themselves.

**Comment by szagi3891** [comment]
This message at compile time will be very helpful.

> This is a breaking change.
Why is this a breaking change ? Until now, generic types have not been generated in the openapi specification (I guess).

**Comment by sunli829** [comment]
> Why is this a breaking change ? Until now, generic types have not been generated in the openapi specification (I guess).

For `poem-openapi`. 🙂

**Comment by sunli829** [comment]
Fixed in `v1.3.28`

**Comment by szagi3891** [comment]
Thank you very much :)
